okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -439,6 +464,69 @@ public void close() {}
     return null;
   }
 
+  private Socket createHttpProxySocket(InetSocketAddress address, InetSocketAddress proxyAddress,
+      String proxyUsername, String proxyPassword) {
+    Status internalStatus = Status.INTERNAL.withDescription("Failed trying to connect with proxy");
+    try {
+      Socket sock = new Socket(proxyAddress.getAddress(), proxyAddress.getPort());
+      sock.setTcpNoDelay(true);
+
+      BufferedSource source = Okio.buffer(Okio.source(sock));
+      BufferedSink sink = Okio.buffer(Okio.sink(sock));
+
+      // Prepare headers and request method line
+      Request proxyRequest = createHttpProxyRequest(address, proxyUsername, proxyPassword);
+      HttpUrl url = proxyRequest.httpUrl();
+      String requestLine = String.format("CONNECT %s:%d HTTP/1.1", url.host(), url.port());
+
+      // Write request to socket
+      sink.writeUtf8(requestLine).writeUtf8("\r\n");
+      for (int i = 0, size = proxyRequest.headers().size(); i < size; i++) {
+        sink.writeUtf8(proxyRequest.headers().name(i))
+            .writeUtf8(": ")
+            .writeUtf8(proxyRequest.headers().value(i))
+            .writeUtf8("\r\n");
+      }
+      sink.writeUtf8("\r\n");
+      // Flush buffer (flushes socket and sends request)
+      sink.flush();
+
+      // Read response, parse headers, check if 200 OK was returned
+      StatusLine statusLine = StatusLine.parse(source.readUtf8LineStrict());
+      if (statusLine.code != 200) {
+        throw new IOException("Response returned from proxy was not 200 OK");
+      }
+      return sock;
+
+    } catch (SocketException e) {
+      // NB(lukaszx0) We'd use mulit-catch, but it's not supported at this language level.
+      throw internalStatus.withCause(e).asRuntimeException();
+    } catch (IOException e) {
+      throw internalStatus.withCause(e).asRuntimeException();
+    }
+  }
+
+  private Request createHttpProxyRequest(InetSocketAddress address, String proxyUsername,
+      String proxyPassword) {
+    HttpUrl tunnelUrl = new HttpUrl.Builder()
+        .scheme("https")
+        .host(address.getHostName())
```

Comment:
As I mentioned in grpc/grpc#8248, while RFC-2817 doesn't explicitly state what the Host: header should be set to, all of the examples in the RFC show it being set to the same value as the argument to the CONNECT request, so I think that's the right behavior.

@jboeuf has floated the idea of using a different value here in cases where we are using both proxies and load-balancing, but he has not yet written a design proposal for this, and I don't think that we should assume anything about this until we have consensus on a design.